bazel/carbon_rules/defs.bzl

```diff
@@ -23,32 +27,91 @@ def carbon_binary(name, srcs):
+            ctx.label.name,
+            src.short_path.removeprefix(ctx.label.package).removesuffix(src.extension),
+        ))
+        objs.append(out)
+        srcs_reordered = [s for s in ctx.files.srcs if s != src] + [src]
+        ctx.actions.run(
+            outputs = [out],
+            inputs = srcs_reordered,
+            executable = toolchain_driver,
+            tools = depset(toolchain_data),
+            arguments = ["compile", "--output=" + out.path] + [s.path for s in srcs_reordered],
+            mnemonic = "CarbonCompile",
+            progress_message = "Compiling " + src.short_path,
         )
 
     # For now, we assume that the prelude doesn't produce any necessary object
     # code, and don't include the .o files for //core/prelude... in the final
     # linked binary.
     #
     # TODO: This will need to be revisited eventually.
-    objs = [s + ".o" for s in srcs]
-    native.genrule(
-        name = name + ".link",
-        tools = [
-            "//toolchain/install:prefix_root/bin/carbon",
-            "//toolchain/install:install_data",
-        ],
-        cmd = "$(execpath //toolchain/install:prefix_root/bin/carbon) link --output=$@ $(SRCS)",
-        srcs = objs,
-        outs = [name],
+    bin = ctx.actions.declare_file(ctx.label.name)
+    ctx.actions.run(
+        outputs = [bin],
+        inputs = objs,
+        executable = toolchain_driver,
+        tools = depset(toolchain_data),
+        arguments = ["link", "--output=" + bin.path] + [o.path for o in objs],
+        mnemonic = "CarbonLink",
+        progress_message = "Linking " + bin.short_path,
+    )
+    return [DefaultInfo(files = depset([bin]))]
+
+carbon_binary_internal = rule(
+    implementation = _carbon_binary_impl,
+    attrs = {
+        # The exec config toolchain driver and data. These will be `None` when
+        # using the target config and populated when using the exec config.
+        "internal_exec_toolchain_data": attr.label(
+            cfg = "exec",
+        ),
+        "internal_exec_toolchain_driver": attr.label(
+            allow_single_file = True,
+            executable = True,
+            cfg = "exec",
+        ),
+
+        # The target config toolchain driver and data. These will be 'None' when
+        # using the exec config and populated when using the target config.
+        "internal_target_toolchain_data": attr.label(
+            cfg = "target",
+        ),
+        "internal_target_toolchain_driver": attr.label(
+            allow_single_file = True,
+            executable = True,
+            cfg = "target",
+        ),
+        "srcs": attr.label_list(allow_files = [".carbon"]),
+    },
+)
+
+def carbon_binary(name, srcs):
+    """Compiles a Carbon binary.
+
+    Args:
+      name: The name of the build target.
+      srcs: List of Carbon source files to compile.
+    """
+    carbon_binary_internal(
+        name = name,
+        srcs = srcs,
+        internal_exec_toolchain_driver = select({
```

Comment:
Out of curiosity, why does this provide exec and target as separate flags instead of the same flag? Is it because you need the difference `cfg` value in the attr? This might be worth a comment explaining the `cfg` nuance.

Had you tried something like:

```
cfg = select({
  # Use the same driver as for the target platform, reducing built artifacts.
  "//bazel/carbon_rules:use_target_config_carbon_rules_config": "target",
  # Use separate host and target drivers.
  "//conditions:default": "exec",
}),
```

(if this works, I'd suggest it to reduce duplication of attrs, but I'm not sure with bazel)

Reply:
Unfortunately doesn't work. Still, would appreciate a comment specific to why the flags are separated.

Reply:
Yeah, sadly this is the only way I could find around the `cfg` setting. Added a comment though, as it is quite mysterious.